File: docker/Dockerfile.java
# Multi-stage Java / Spring Boot production Dockerfile
# Supports: Spring Boot 3.x, Java 21 (LTS), Gradle or Maven

# ─── Stage 1: Build (Maven) ───────────────────────────────────────────
FROM eclipse-temurin:21-jdk-alpine AS builder
WORKDIR /app

# Layer caching: copy POM/Gradle files first
COPY pom.xml ./
COPY .mvn/ .mvn/
COPY mvnw ./
RUN chmod +x mvnw

# Download dependencies (cached layer)
RUN ./mvnw dependency:go-offline -B

# Copy source and build
COPY src ./src
RUN ./mvnw package -DskipTests -B --no-transfer-progress

# Extract Spring Boot layers for optimized Docker caching
RUN java -Djarmode=layertools -jar target/*.jar extract --destination /extracted

# ─── Stage 2: JRE Runtime ─────────────────────────────────────────────
FROM eclipse-temurin:21-jre-alpine AS runner
WORKDIR /app

ENV JAVA_OPTS="-XX:+UseContainerSupport -XX:MaxRAMPercentage=75 -XX:+ExitOnOutOfMemoryError -Djava.security.egd=file:/dev/./urandom"
ENV SPRING_PROFILES_ACTIVE=production
ENV PORT=8080

# Security: non-root user
RUN addgroup --system --gid 1001 javagroup \
    && adduser --system --uid 1001 --ingroup javagroup appuser

# Spring Boot layered JAR for fast startup and optimized caching
COPY --from=builder --chown=appuser:javagroup /extracted/dependencies/ ./
COPY --from=builder --chown=appuser:javagroup /extracted/spring-boot-loader/ ./
COPY --from=builder --chown=appuser:javagroup /extracted/snapshot-dependencies/ ./
COPY --from=builder --chown=appuser:javagroup /extracted/application/ ./

LABEL org.opencontainers.image.title="Spring Boot Application" \
      org.opencontainers.image.vendor="Your Org" \
      org.opencontainers.image.licenses="Apache-2.0"

USER appuser

EXPOSE 8080

HEALTHCHECK --interval=30s --timeout=10s --start-period=60s --retries=3 \
  CMD wget --quiet --tries=1 --spider http://localhost:8080/actuator/health || exit 1

ENTRYPOINT ["sh", "-c", "java $JAVA_OPTS org.springframework.boot.loader.launch.JarLauncher"]

File: docker/Dockerfile.go
# Multi-stage Go production Dockerfile
# Produces a minimal scratch image (< 20MB)

# ─── Stage 1: Build ───────────────────────────────────────────────────
FROM golang:1.22-alpine AS builder
WORKDIR /app

# Install CA certs and timezone data for scratch image
RUN apk add --no-cache ca-certificates tzdata

# Dependency caching
COPY go.mod go.sum ./
RUN go mod download && go mod verify

# Build static binary
COPY . .
RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 \
    go build \
    -ldflags="-w -s -X main.version=$(cat VERSION 2>/dev/null || echo dev) -X main.buildTime=$(date -u +%Y-%m-%dT%H:%M:%SZ)" \
    -trimpath \
    -o /server \
    ./cmd/server

# ─── Stage 2: Production (scratch) ───────────────────────────────────
FROM scratch AS runner

# Copy essentials from builder
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
COPY --from=builder /usr/share/zoneinfo /usr/share/zoneinfo
COPY --from=builder /server /server

# Non-root UID (numeric, scratch has no useradd)
USER 65532:65532

LABEL org.opencontainers.image.title="Go Application" \
      org.opencontainers.image.vendor="Your Org" \
      org.opencontainers.image.licenses="Apache-2.0"

EXPOSE 8080

HEALTHCHECK --interval=30s --timeout=5s --start-period=10s --retries=3 \
  CMD ["/server", "health"]

ENTRYPOINT ["/server"]
